src/main/java/com/strandls/document/service/Impl/DocumentServiceImpl.java

```diff
@@ -1487,4 +1504,169 @@ public Boolean documentDownloadLog(HttpServletRequest request, DownloadLogData d
+						saveScientificNamesInTable(sortedNames, documentId);
+					}
+
+				}
+
+			} catch (Exception e) {
+				logger.error(e.getMessage());
+			}
+
+		} catch (URISyntaxException e1) {
+			logger.error(e1.getMessage());
+		}
+
+		return gnfinderresponse;
+	}
+
+	private List<GnFinderResponseNames> sortScientificNamesOnFrequency(List<GnFinderResponseNames> names) {
+		Collections.sort(names, new Comparator<GnFinderResponseNames>() {
+			public int compare(GnFinderResponseNames name1, GnFinderResponseNames name2) {
+				return ((Integer) name2.getFrequency()).compareTo(name1.getFrequency());
+			}
+		});
+
+		return names;
+	}
+
+	private void saveScientificNamesInTable(List<GnFinderResponseNames> names, Long documentId) {
+		int order = 1;
+		for (GnFinderResponseNames name : names) {
+			DocSciName nameTosave = new DocSciName();
+			nameTosave.setCanonicalForm(name.getName());
+			nameTosave.setDocumentId(documentId);
+			nameTosave.setFrequency(name.getFrequency());
+			nameTosave.setIsDeleted(false);
+			nameTosave.setOffsetValues(name.getOffSet());
+			nameTosave.setScientificName(name.getName());
+			nameTosave.setTaxonConceptId(name.getTaxonId());
+			nameTosave.setVersion(0L);
+			nameTosave.setDisplayOrder(order);
+			nameTosave.setPrimaryName(0);
+			order++;
+			docSciNameDao.save(nameTosave);
+		}
+	}
+
+	private String concatValues(Object value1, Object value2) {
+		StringBuilder stringBuilder = new StringBuilder();
+		stringBuilder.append("[").append(value1).append(",").append(value2).append("]");
+		return stringBuilder.toString();
+	}
+
+	private GNFinderResponseMap mergeCommonObjects(GNFinderResponseMap response) {
+		HashMap<String, GnFinderResponseNames> scientificNameMap = new HashMap<String, GnFinderResponseNames>();
+		List<GnFinderResponseNames> names = response.getNames();
+
+		for (GnFinderResponseNames name : names) {
+			String scientificNameKey = name.getName();
+			String nameOffset = concatValues(name.getStart(), name.getEnd());
+
+			if (scientificNameMap.containsKey(scientificNameKey) == true) {
+				StringBuilder stringBuilder = new StringBuilder();
+				GnFinderResponseNames existingName = scientificNameMap.get(scientificNameKey);
+				String offSet = existingName.getOffSet();
+				existingName.setOffSet(stringBuilder.append(offSet).append(",").append(nameOffset).toString());
+				Integer prevFreq = existingName.getFrequency();
+				existingName.setFrequency(prevFreq + 1);
+				scientificNameMap.replace(scientificNameKey, existingName);
+			} else {
+				name.setOffSet(nameOffset);
+				name.setFrequency(1);
+				scientificNameMap.put(scientificNameKey, name);
+			}
+		}
+		ArrayList<GnFinderResponseNames> ans = new ArrayList<GnFinderResponseNames>(scientificNameMap.values());
+		response.setNames(ans);
+		return response;
+	}
+
+	public List<DocSciName> getNamesByDocumentId(Long documentId, int offset) {
+
+		try {
+
+			List<DocSciName> response = docSciNameDao.findByDocId(documentId);
```

Comment:
use offset directly in the dao class instead of writing all thee unnecessary logic here, also if suppose offset is null, then respond back all the documents

Reply:
Corrected.